src/plugins/es_ui_shared/public/request/use_request.ts

```diff
@@ -33,114 +37,129 @@ export interface UseRequestResponse<D = any, E = Error> {
+  const queryToString = _query ? JSON.stringify(_query) : undefined;
 
-  // Tied to every render and bound to each request.
-  let isOutdatedRequest = false;
+  const requestBody = useMemo(() => {
+    return {
+      path,
+      method,
+      query: queryToString ? JSON.parse(queryToString) : undefined,
+      body: bodyToString ? JSON.parse(bodyToString) : undefined,
+    };
+  }, [path, method, queryToString, bodyToString]);
 
-  scheduleRequestRef.current = () => {
-    // Clear current interval
+  const cleanUpPollInterval = useCallback(() => {
     if (pollIntervalIdRef.current) {
       clearTimeout(pollIntervalIdRef.current);
     }
+  }, []);
 
-    // Set new interval
-    if (pollIntervalMsRef.current) {
-      pollIntervalIdRef.current = setTimeout(sendRequestRef.current!, pollIntervalMsRef.current);
-    }
-  };
+  const sendRequest = useCallback(async () => {
+    cleanUpPollInterval();
 
-  sendRequestRef.current = async () => {
     // We don't clear error or data, so it's up to the consumer to decide whether to display the
     // "old" error/data or loading state when a new request is in-flight.
     setIsLoading(true);
 
-    const requestBody = {
-      path,
-      method,
-      query,
-      body,
-    };
-
-    const response = await sendRequest<D, E>(httpClient, requestBody);
+    const response = await _sendRequest<D, E>(httpClient, requestBody);
     const { data: serializedResponseData, error: responseError } = response;
 
-    // If an outdated request has resolved, ignore its outdated data.
-    if (isOutdatedRequest) {
+    if (isMounted.current === false) {
       return { data: null, error: null };
     }
 
+    setIsLoading(false);
+    setTotalRequests((prev) => prev + 1);
+
     setError(responseError);
     // If there's an error, keep the data from the last request in case it's still useful to the user.
     if (!responseError) {
-      const responseData = deserializer(serializedResponseData);
+      const responseData = deserializer
+        ? deserializer(serializedResponseData)
+        : serializedResponseData;
       setData(responseData);
     }
-    setIsLoading(false);
-    setIsInitialRequest(false);
-
-    // If we're on an interval, we need to schedule the next request. This also allows us to reset
-    // the interval if the user has manually requested the data, to avoid doubled-up requests.
-    scheduleRequestRef.current!();
 
     return { data: serializedResponseData, error: responseError };
-  };
+  }, [requestBody, httpClient, deserializer, cleanUpPollInterval]);
+
+  const scheduleRequest = useCallback(() => {
+    cleanUpPollInterval();
+
+    if (pollIntervalMs) {
+      pollIntervalIdRef.current = setTimeout(sendRequest, pollIntervalMs);
+    }
+  }, [pollIntervalMs, sendRequest, cleanUpPollInterval]);
+
+  useEffect(() => {
+    sendRequest();
+  }, [sendRequest]);
 
+  // Whenever the scheduleRequest() changes (which changes when the pollIntervalMs changes)
```

Comment:
From what we discussed yesterday, I've been thinking and it's the first `useEffect` that we could remove.

So the end result is

```
// Send the request on component mount
// + anytime the deps of sendRequest() change
useEffect(() => {
  sendRequest();
}, [sendRequest]);

// Schedule a new request each time a request fulfills ("totalRequests" increment)
// + anytime the deps of scheduleRequest() change
useEffect(() => {
  if (isMounted.current === false) {
    // Don't schedule on component mount
    return;
  }


  scheduleRequest();

  return () => {
    console.log('Effect clean up!');
    cleanUpPollInterval();
  };
}, [totalRequests, scheduleRequest, cleanUpPollInterval]);

useEffect(() => {
  isMounted.current = true;

  return () => {
    isMounted.current = false;
    // When the component unmounts, clear any existing interval.
    cleanUpPollInterval();
  };
}, [cleanUpPollInterval]);
```

Reply:
And this revealed a bug in the test for pollInterval. We need to wait for twice the amount of time to check for the `sendRequestSpy` to have been called. We wait once for the pollIntervalMs to pass and once for the `setTimeout` time to pass (defined in the mock HTTP client post() handler). @cjcenizal 

So the test becomes

```js
describe('pollIntervalMs', () => {
  it('sends another request after the specified time has elapsed', async () => {
    const { setupSuccessRequest, advanceTime, getSendRequestSpy } = helpers;
    setupSuccessRequest({ pollIntervalMs: REQUEST_TIME });

    await advanceTime(REQUEST_TIME);
    expect(getSendRequestSpy().callCount).toBe(1);

    // We need to advance (1) the pollIntervalMs and (2) the request time
    await advanceTime(REQUEST_TIME * 2);
    expect(getSendRequestSpy().callCount).toBe(2);

    await advanceTime(REQUEST_TIME * 2);
    expect(getSendRequestSpy().callCount).toBe(3);
  });
});
```